Adds invalid char validators for rollup tags and target names

m-sandusky opened the pull request:
Adds to validator_options as well. Wasn't sure whether to call RollupTargetName or RollupMetricName

xichen2020 commented on rules/validator.go:
```diff
@@ -125,6 +125,11 @@ func (v *validator) validateRollupRules(rrv map[string]*RollupRuleView) error {
 		}
 
 		for _, target := range view.Targets {
+			// Validate that rollup target name is valid
```
nit: comment should be a complete sentence and end w/ a period.

xichen2020 replied:
Might also want to extract this logic into a `valiateRollupMetricName` method, in which we can also check if the metric name is empty etc.

xichen2020 commented on rules/validator.go:
```diff
@@ -125,6 +125,11 @@ func (v *validator) validateRollupRules(rrv map[string]*RollupRuleView) error {
 
 		for _, target := range view.Targets {
+			// Validate that rollup target name is valid
+			if err := validateChars(target.Name, v.opts.RollupTargetNameInvalidChars()); err != nil {
```
nit: assign `invalidChars := v.opts. RollupTargetNameInvalidChars` outside the for loop and reuse it here? You could it just cache it inside the validator to avoid function calls completely.

xichen2020 commented on rules/validator.go:
```diff
@@ -125,6 +125,11 @@ func (v *validator) validateRollupRules(rrv map[string]*RollupRuleView) error {
 		for _, target := range view.Targets {
+			// Validate that rollup target name is valid
+			if err := validateChars(target.Name, v.opts.RollupTargetNameInvalidChars()); err != nil {
+				return fmt.Errorf("rollup rule %s has an invalid rollup target name %s: %s", view.Name, target.Name, err.Error())
```
Perhaps print the error as `fmt.Errorf(...: %v, ..., err)`?

xichen2020 commented on rules/validator.go:
```diff
@@ -164,11 +169,17 @@ func (v *validator) validateFilters(f map[string]string) error {
+	// Validate that all tags have valid characters
+	invalidChars := v.opts.RollupTagInvalidChars()
+	for _, tag := range tags {
+		fmt.Println("in here")
```
Remove debugging println?

m-sandusky replied:
Yikes, thanks for catching that!

xichen2020 commented on rules/validator.go:
```diff
@@ -164,11 +169,17 @@ func (v *validator) validateFilters(f map[string]string) error {
+	for _, tag := range tags {
+		fmt.Println("in here")
+		if err := validateChars(tag, invalidChars); err != nil {
+			return fmt.Errorf("invalid rollup tag %s for rollup rule %s : %s", tag, ruleName, err.Error())
```
Perhaps make the error message consistent w/ others, e.g., something along the lines of `fmt.Errorf("rollup rule %s has invalid rollup tag %s: %v", ruleName, tag, err)`?

xichen2020 commented on rules/validator.go:
```diff
@@ -164,11 +169,17 @@ func (v *validator) validateFilters(f map[string]string) error {
-	requiredTags := v.opts.RequiredRollupTags()
-	if len(requiredTags) == 0 {
-		return nil
+	// Validate that all tags have valid characters
```
The tags passed in here are tag names rather than tag values btw.

Also comment should end w/ a period.

xichen2020 commented on rules/validator.go:
```diff
@@ -164,11 +169,17 @@ func (v *validator) validateFilters(f map[string]string) error {
 	}
+
 	// Validating the list of rollup tags in the rule contain all required tags.
+	requiredTags := v.opts.RequiredRollupTags()
```
Perhaps still shortcut it by checking `len(requiredTags) == 0` and returning early if so to avoid allocating the map below?

m-sandusky replied:
Yeah, good point. I also enclosed the invalid chars logic in an if len(invalidChars) > 0 clause - might be a little ugly but avoids unnecessary iteration through the tags. Let me know if you there's a better way.

xichen2020 commented on rules/validator.go:
```diff
@@ -178,6 +189,25 @@ func (v *validator) validateRollupTags(ruleName string, tags []string) error {
+		return nil
+	}
+
+	// Validate that given string doesn't contain an invalid character
```
Comment should end w/ a period.

xichen2020 commented on rules/validator.go:
```diff
@@ -178,6 +189,25 @@ func (v *validator) validateRollupTags(ruleName string, tags []string) error {
+	}
+
+	// Validate that given string doesn't contain an invalid character
+	chars := make(map[rune]struct{}, len(invalidChars))
```
It might actually be faster if we simply iterate over the slice.

If not, I'd actually just construct the map once and store it in the validator and reuse it here to avoid reconstructing it on every validation call.

m-sandusky replied:
Yeah, makes sense - I'll construct the mapping in the validator and just past that along here.

xichen2020 commented on rules/validator.go:
```diff
@@ -178,6 +189,25 @@ func (v *validator) validateRollupTags(ruleName string, tags []string) error {
+	}
+	for _, char := range str {
+		if _, exists := chars[char]; exists {
+			return fmt.Errorf("%d is an invalid character", char)
```
nit: perhaps it should be `"%s contains invalid character %v", char`? Also would be good to mention the rule name here as well to give more context around the error.

m-sandusky replied:
Sgtm regarding "%s contains invalid character %v" . Regarding the rule name, since this validator is a simple util that could be used anywhere, I wanted to avoid passing in values that contain knowledge of specific things like rule name. The validators that do use util would append this error to their own message. Ie. this function validates that if a string contains invalid chars and bubbles that up. Happy to change if you think it would be better to add rule name however.

xichen2020 replied:
Yeah fair point, my concern was that the error message will be bubbled back to the frontend, and when the end user sees this message, having a rule name might help them identify the root cause of the issue more quickly. But I see that the caller reformats the error to include the rule name, works for me.

xichen2020 commented:
`RollupMetricName` is probably more accurate btw.

xichen2020 commented on rules/validator_options.go:
```diff
@@ -56,6 +56,18 @@ type ValidatorOptions interface {
 	RequiredRollupTags() []string
 
+	// SetRollupTagInvalidChars sets the list of invalid chars for a rollup metric tag.
+	SetRollupTagInvalidChars([]rune) ValidatorOptions
```
I'd just call this `SetTagNameInvalidChars` since they apply to the tags of all metrics. We could use them to validate the tags in the tag filters for both mapping rules and rollup rules.

m-sandusky replied:
Yeah that sounds much better, done.

xichen2020 commented on rules/validator_options.go:
```diff
@@ -56,6 +56,18 @@ type ValidatorOptions interface {
+	RollupTagInvalidChars() []rune
+
+	// SetRollupTargetNameInvalidChars sets the list of invalid chars for a rollup target name.
+	SetRollupTargetNameInvalidChars([]rune) ValidatorOptions
```
Similarly I'd just call this `SetMetricNameInvalidChars`.

m-sandusky commented:
@xichen2020 Thanks for the review! I've addressed your comments. Will add validation for filter names in a separate pull request to avoid this one getting too large. The major thing I changed was having the validator options transform the input invalidChars into a map[rune]struct{} - which I'm not totally sure is the best way in terms of code clarity so would appreciate feedback there!

xichen2020 commented on rules/validator.go:
```diff
@@ -178,6 +200,35 @@ func (v *validator) validateRollupTags(ruleName string, tags []string) error {
+}
+
+func (v *validator) validateRollupMetricName(metricName, ruleName string) error {
+	// Validate that rollup metric name is not empty
```
nit: add a period at the end.

xichen2020 commented on rules/validator.go:
```diff
@@ -178,6 +200,35 @@ func (v *validator) validateRollupTags(ruleName string, tags []string) error {
+func (v *validator) validateRollupMetricName(metricName, ruleName string) error {
+	// Validate that rollup metric name is not empty
+	if metricName == "" {
+		return errors.New("rollup rule %s has an empty rollup metric name")
```
Probably want `fmt.Errorf(..., ruleName)` here?

m-sandusky replied:
woop good catch - thanks

xichen2020 commented:
For the `map[rune]struct{}`, it's probably ok. Alternatively, you could define `ContainsInvalidCharsForName` and `ContainsInvalidCharsForTag` on the options so you can remove the two `*InvalidChars` getters and hide the implementation details.

xichen2020 commented on rules/validator_options.go:
```diff
@@ -56,6 +58,20 @@ type ValidatorOptions interface {
 	RequiredRollupTags() []string
 
+	// SetTagNameInvalidChars sets the list of invalid chars for a tag name.
+	SetTagNameInvalidChars([]rune) ValidatorOptions
```
nit: convention is to put `value` as the var name.

xichen2020 commented on rules/validator_options.go:
```diff
@@ -56,6 +58,20 @@ type ValidatorOptions interface {
+	ContainsInvalidCharactersForTagName(tagName string) (bool, error)
+
+	// SetMetricNameInvalidChars sets the list of invalid chars for a metric name.
+	SetMetricNameInvalidChars([]rune) ValidatorOptions
```
Same as above.

m-sandusky commented on rules/validator.go:
```diff
@@ -154,7 +159,12 @@ func (v *validator) validateRollupRules(rrv map[string]*RollupRuleView) error {
 }
 
 func (v *validator) validateFilters(f map[string]string) error {
```
Yup sounds good

m-sandusky commented on rules/validator_options.go:
```diff
@@ -130,6 +148,40 @@ func (o *validatorOptions) RequiredRollupTags() []string {
+	return &opts
+}
+
+func (o *validatorOptions) ContainsInvalidCharactersForTagName(tagName string) (bool, error) {
```
Yeah, good point here. I felt that this was a little awkward, but the thinking was a function with a name like ContainsX would be a boolean return value. I'll change it to just return error.

xichen2020 replied:
That's probably fine, or perhaps you can rename it to `CheckInvalidCharsForTagName`?

xichen2020 commented:
Seems some tests are failing?

xichen2020 commented on rules/validator_options.go:
```diff
@@ -173,3 +219,17 @@ func toAggregationTypeSet(aggTypes policy.AggregationTypes) map[policy.Aggregati
+	// Validate that given string doesn't contain an invalid character.
+	for _, char := range str {
+		if _, exists := invalidChars[char]; exists {
+			return fmt.Errorf("%s contains invalid character %v", str, char)
```
Probably want to use `fmt.Errorf("%s contains invalid character %s", str, strconv.QuoteRune(char))`? Otherwise the error message will contain the numerical value of the character instead of the human-readable string.
